migration/migration.go

```diff
@@ -56,47 +73,69 @@ var migrations = []Migration{
 	/*
-		Schema version of the targetDB determines which set of migrations need to be applied to the database.
+		Schema metadata of the targetDB determines which set of migrations need to be applied to the database.
 		After a migration is successfully executed, which may update the database, the schema version is incremented
 		by 1 by this loop.
 
 		For example;
 
 		Assume an empty DB, which has a schema version of 0. So we start applying the migrations from the
 		very first one in the list and increment the schema version as we go. After the loop exists we
-		end up with a database which's schema version is len(migrations).
+		end up with a database which's schema version is len(migrations)-1 and done set to true.
 
 		After running that Juno version for a while, if the user updates its Juno version which adds new
 		migrations to the list, MigrateIfNeeded will skip the already applied migrations and only apply the
 		new ones. It will be able to do this since the schema version it reads from the database will be
 		non-zero and that is what we use to initialise the i loop variable.
 	*/
-	version, err := SchemaVersion(targetDB)
+	metadata, err := SchemaMetadata(targetDB)
 	if err != nil {
 		return err
 	}
 
-	for i := version; i < uint64(len(migrations)); i++ {
+	for i := metadata.Version; i < uint64(len(migrations)); i++ {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
 		log.Infow("Applying database migration", "stage", fmt.Sprintf("%d/%d", i+1, len(migrations)))
 		migration := migrations[i]
-		migration.Before()
-		for {
+		err := migration.Before(metadata.IntermediateState)
+		if err != nil {
+			return err
+		}
+		var intermediateState []byte
+		for x := uint64(0); ; x++ {
 			var migrationErr error
 			if dbErr := targetDB.Update(func(txn db.Transaction) error {
-				migrationErr = migration.Migrate(txn, network)
+				intermediateState, migrationErr = migration.Migrate(ctx, txn, network)
+				// if migration returns any state then we should keep it as latest sub-stage.
+				if intermediateState != nil {
+					metadata.IntermediateState = intermediateState
+				}
 				if migrationErr != nil {
+					// save the current progress.
+					if errors.Is(migrationErr, ctx.Err()) && metadata.IntermediateState != nil {
+						return updateSchemaMetadata(txn, metadata)
+					}
 					if errors.Is(migrationErr, ErrCallWithNewTransaction) {
 						return nil // Run the migration again with a new transaction.
 					}
 					return migrationErr
 				}
-
-				// Migration successful. Bump the version.
-				var versionBytes [8]byte
-				binary.BigEndian.PutUint64(versionBytes[:], i+1)
-				return txn.Set(db.SchemaVersion.Key(), versionBytes[:])
+				// Migration successful. Update metadata with new version and none intermediate
+				metadata.Version++
+				metadata.IntermediateState = nil
+				return updateSchemaMetadata(txn, metadata)
```

Comment:
I think we can replace this with:
```go
metadata.IntermediateState, migrationErr = migration.Migrate(ctx, txn, network)
switch migrationErr {
case ctx.Err():
case nil:
case ErrCallWithNewTransaction:
	return nil // Run migration again with new transaction.
default:
	return migrationErr
}
if metadata.IntermediateState == nil {
	metadata.Version++
}
return updateSchemaMetadata(txn, metadata)
```

Reply:
Does the switch match the same way as `errors.Is` ? There are some migrations that return wrapped errors, eg. `db.CloseAndWrapOnError(iterator.Close, ErrCallWithNewTransaction)`

Reply:
@joshklop What do you think? I don't know if the behavior change is intentional :sweat_smile:

Reply:
I haven't looked into this very deeply, but if there is a behavior change, then one of the implementations should cause a test to fail. Right now, both implementations pass all tests, which indicates we may be missing some edge cases in tests. Please correct me if I'm missing something.